libr/core/anal_tp.c

```diff
@@ -155,6 +155,80 @@ static ut64 get_addr(Sdb *trace, const char *regname, int idx) {
 	return r_num_math (NULL, sdb_const_get (trace, query, 0));
 }
 
+static int cond_invert (int cond) {
+	int res = 0;
+	switch (cond) {
+	case R_ANAL_COND_LE:
+		res = R_ANAL_COND_GT;
+		break;
+	case R_ANAL_COND_LT:
+		res = R_ANAL_COND_GE;
+		break;
+	case R_ANAL_COND_GE:
+		res = R_ANAL_COND_LT;
+		break;
+	case R_ANAL_COND_GT:
+		res = R_ANAL_COND_LE;
+		break;
+	}
+	return res;
+}
+
+#define RKEY(a,k,d) sdb_fmt ("var.range.0x%"PFMT64x ".%c.%d", a, k, d)
+#define ADB a->sdb_fcns
+
+static void var_add_range (RAnal *a, RAnalVar *var, int cond, ut64 val) {
+	const char *key = RKEY (var->addr, var->kind, var->delta);
+	sdb_array_append_num(ADB, key, cond, 0);
+	sdb_array_append_num(ADB, key, val, 0);
+}
+
+R_API char *var_get_constraint (RAnal *a, RAnalVar *var) {
+	char res[256] = {0};
+	const char *key = RKEY (var->addr, var->kind, var->delta);
+	int i, n = sdb_array_length(ADB, key);
+	int len = sizeof (res) - 1;
+	bool low = false, high = false;
+
+	if (n < 2) {
+		return NULL;
+	}
+	for (i = 0; i < n; i += 2) {
+		ut64 cond = sdb_array_get_num (ADB, key, i, 0);
+		ut64 val = sdb_array_get_num (ADB, key, i + 1, 0);
+		switch (cond) {
+		case R_ANAL_COND_LE:
+			if (high) {
+				strncat (res, " && ", len);
+			}
+			strncat (res, sdb_fmt ("<= 0x%"PFMT64x "", val), len);
+			low = true;
+			break;
+		case R_ANAL_COND_LT:
+			if (high) {
+				strncat (res, " && ", len);
+			}
+			strncat (res, sdb_fmt ("< 0x%"PFMT64x "", val), len);
```

Comment:
Dont use strncat. Better use strbuf if u have to append many strings. Or rstring